File: laxmi-agro-backend/src/utils/phoneValidation.js
const INDIAN_MOBILE_PATTERN = /^[6-9]\d{9}$/;

function hasRepeatedDigits(phone) {
  return /^(\d)\1{9}$/.test(phone);
}

function hasSequentialDigits(phone) {
  if (!/^\d{10}$/.test(phone)) return false;

  const digits = [...phone].map(Number);
  const isAscending = digits.every((digit, index) => (
    index === 0 || digit === (digits[index - 1] + 1) % 10
  ));
  const isDescending = digits.every((digit, index) => (
    index === 0 || digit === (digits[index - 1] + 9) % 10
  ));

  return isAscending || isDescending;
}

function isRegistrationPhoneValid(phone) {
  return typeof phone === 'string'
    && INDIAN_MOBILE_PATTERN.test(phone)
    && !hasRepeatedDigits(phone)
    && !hasSequentialDigits(phone);
}

module.exports = {
  INDIAN_MOBILE_PATTERN,
  hasRepeatedDigits,
  hasSequentialDigits,
  isRegistrationPhoneValid,
};

File: laxmi-agro-app/lib/core/utils/phone_validation.dart
class PhoneValidation {
  static final RegExp _indianMobilePattern = RegExp(r'^[6-9]\d{9}$');
  static final RegExp _repeatedDigitsPattern = RegExp(r'^(\d)\1{9}$');

  static bool isValidRegistrationPhone(String phone) {
    return _indianMobilePattern.hasMatch(phone) &&
        !_repeatedDigitsPattern.hasMatch(phone) &&
        !_hasSequentialDigits(phone);
  }

  static String? registrationError(String phone) {
    if (!_indianMobilePattern.hasMatch(phone)) {
      return 'Enter a valid 10-digit Indian mobile number starting with 6, 7, 8, or 9.';
    }

    if (_repeatedDigitsPattern.hasMatch(phone) || _hasSequentialDigits(phone)) {
      return 'Enter a real mobile number, not a repeated or sequential number.';
    }

    return null;
  }

  static bool _hasSequentialDigits(String phone) {
    if (!RegExp(r'^\d{10}$').hasMatch(phone)) return false;

    final digits = phone.codeUnits.map((codeUnit) => codeUnit - 48).toList();
    final isAscending = List.generate(digits.length - 1, (index) => index).every(
      (index) => digits[index + 1] == (digits[index] + 1) % 10,
    );
    final isDescending = List.generate(digits.length - 1, (index) => index).every(
      (index) => digits[index + 1] == (digits[index] + 9) % 10,
    );

    return isAscending || isDescending;
  }
}
